### DISCUSSION TITLE
Capfs with deprecated calling convention

### DISCUSSION THREAD

**Discussion by minad** [company-mode/company-mode#1278] [Q&A]
Hi @dgutov,

I would like to ask about the current status of misbehaving Capfs. I've noticed a few times that pcomplete-completions-at-point sometimes misbehaves in Eshell.  However it does not always misbehave, only sometimes, for example after a " or a '. Since the function mostly behaves correctly I don't want to disable/remove it entirely as proposed in https://github.com/company-mode/company-mode/issues/409 and https://github.com/company-mode/company-mode/issues/974.

Are you aware of other misbehaving Capfs or is Pcomplete the only offender?

Anyways, I have a workaround in my Cape package which "fixes" the issue. It is of course no long term solution, only a stop gap measure. See https://github.com/minad/cape#repair-misbehaving-capfs, `cape-repair-misbehaving-capf`.

Thanks!

**Comment by minad** [comment]
There is also this upstream bug report: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=40651

**Comment by dgutov** [comment]
Hi Daniel,

Also see https://github.com/company-mode/company-mode/discussions/1276 and https://debbugs.gnu.org/cgi/bugreport.cgi?bug=50470 mentioned there which also has a workaround in comments.

And being able to avoid the problem with quotes in Eshell would be really nice.

pcomplete is the only reasonably common offender that I'm aware of, and I've been kind of hoping someone would just go and fix it in there (since apparently this only happens in some branches of its execution, and otherwise the bridge works as a reasonable CAPF). But maybe a workaround is more realistic for now.

Though if we go with the latter, the question is where to apply it. As Stefan probably meant to say, maybe `completion--capf-wrapper` is the place to do it? Though frontends would probably carry the workaround too for a while.

**Comment by minad** [comment]
Thanks for more background and the links! I am not sure if the capf wrapper is the right place for a fix. If the bug is fixed upstream it should be fixed directly in pcomplete.el. For Corfu/Company I was more concerned about fixes in the packages such that "old" Emacsen like 28 profit too!

**Reply by dgutov** [reply]
> If the bug is fixed upstream it should be fixed directly in pcomplete.el.

If by any chance you have both the time and the will, a fix there would be perfect. I've made the attempt to investigate this a couple of times in the past, but didn't manage to advance too far.

**Reply by minad** [reply]
Sorry. No, I am not sending patches to upstream. See https://debbugs.gnu.org/cgi/bugreport.cgi?bug=52459 for a recent discussion I had with Eli. Contributing to upstream is a pain and a waste of time, even if one only makes a proposal. And I would say it is not only my fault.

**Reply by dgutov** [reply]
OK, that's unfortunate.

**Comment by minad** [comment]
Which workaround are you referring to? In my cape repair wrapper I manually restore the point afterwards but it seems like a much better idea to switch the buffer to read only. 

This is indeed an easy enough fix to even put it in the capf wrapper. It just enforces the rules once and forever 😆 

I will try this in Corfu and the report back!

**Reply by minad** [reply]
@dgutov I adopted @jakanakaevangeli's workaround in my `corfu--capf-wrapper`:

https://github.com/minad/corfu/blob/160bd53db56f279d87377a8068aab754ec4b5d0c/corfu.el#L1153-L1183

So far it seems to work well. I suggest you do the same in Company. A similar patch can also be used for minibuffer.el as @monnier proposed in bug#50470.

**Reply by dgutov** [reply]
Yeah, sounds like a good idea, thanks. Will you propose that patch to the core, or should I?

**Reply by dgutov** [reply]
> I suggest you do the same in Company.

And done!

**Reply by minad** [reply]
Hi Dmitry!

In Corfu my approach changed recently. I replace the capf wrapper in its entirety:

https://github.com/minad/corfu/blob/c8e6607c90a89ff19062cd37afc17e8bbb86aba3/corfu.el#L1220

Furthermore broken Capfs must be purified individually:

https://github.com/minad/cape/blob/ee826d70fdd2e8324a4af1a427ea9ce54f5e32e6/cape.el#L1081

See also:

https://github.com/minad/corfu#completing-with-corfu-in-the-eshell-or-shell

**Reply by minad** [reply]
There are multiple reasons for the change:

- pcomplete is broken in more ways. It also requires silencing (catching errors and messages). Therefore I figured it is better to fix all these bugs individually via separate targeted cape wrappers.
- pcomplete is the only capf which is broken so severely. Using a global fix seemed misplaced.
- The original fix (the fix you adopted) is unfortunately incomplete. The capf wrapper returns a pair of capf function and capf result. The returned capf function is not wrapped, which leads to problems when the capf predicate is called! However with cape-wrap-purify all calls to the capf are protected.
- replacing the capf wrapper entirely was necessary since the minibuffer.el capf wrapper is somewhat broken beyond repair. It was easier to replace it right away with an override advice in Corfu. This way I get exactly the behavior I want.

**Reply by minad** [reply]
I just checked the company-capf code and it seems that you never call the capf again (the function returned as car of the cons pair result returned by the capf wrapper) as a predicate to check if the completion position changed. Therefore I think that your current fix is sufficient in Company.

**Reply by dgutov** [reply]
Hi Daniel!

Thanks for the review.

I considered replacing the wrapper entirely, but the main additional value, it seemed, was in better `:exclusive` support (at the expense of performance, in those cases), so I decided against it for now.

And indeed, `fun` is not used after the completion data is produced. So that seems fine.

**Reply by minad** [reply]
Yes, the main advantage is the (as I claim) more correct behavior for :exclusive. In the case of Corfu there is no significant performance penalty since I also moved the prefix length check to the wrapper. However note that Stefan Monnier disagrees with me with respect to what is correct behavior here. Stefan deliberately used only prefix matching to give lower priority Capfs a higher chance to run. Stefan and I had some longer discussion about this via mail. I still stick with my opinion that corfu--capf-wrapper is "more correct" ;) However one possible way out would be the ability to merge Capfs as is done in cape-super-capf (or also via the Company backend merging functionality). This comes with its own set of problems though, e.g., if the completion table uses boundaries or the completion prefixes are different.

**Reply by dgutov** [reply]
I think the consensus here is `:exclusive` is broken and shouldn't be used anyway.

Your solution is arguably more correct, but the performance difference is real because `completion-try-completion` has to go through the completion styles and also (currently, though not by strict necessity) creates the full list of matches for the given prefix. The latter might be cached (if the completion table does that), but the former is not.

**Reply by minad** [reply]
> I think the consensus here is :exclusive is broken and shouldn't be used anyway.

I disagree with that. I think with my fix it behaves as expected. I am a happy user of non-exclusive! However I agree that it has issues and needs fine tuning of the Capf list and the completion styles. If a completion style is too flexible, the higher priority Capfs match too often.

But overall in restricted scenarios everything works well. The situation is not worse than with multiple Company backends. I use non-exclusive Capfs successfully with my Cape package and my Tempel package. In both packages the non-exclusive Capfs are "auxillary" and stay out of the way. They get only activated if the main Capf doesn't take over or if the input matches the template name exactly as in tempel-expand.

> Your solution is arguably more correct, but the performance difference is real because completion-try-completion has to go through the completion styles and also (currently, though not by strict necessity) creates the full list of matches for the given prefix. The latter might be cached (if the completion table does that), but the former is not.

Sure that's an issue. However if you use a well-engineered completion style like Orderless the issue is not severe. See https://github.com/oantolin/orderless/blob/f2c78c4a6059c5f892e48a3887d4368a547515ff/orderless.el#L415 for the trickery which prevents the allocation of the list of completions.

Also even if your completion style is not as optimized as orderless the problem is not severe since we require a minimum prefix. This already ensures that the list is filtered to a small number of candidates. That's the reason why I had to move the prefix check to the caof wrapper.

Allocating the conses for the small list is not that costly. (It is much more costly to allocate the string itself as is done by completion-all-completions. In Vertico and Corfu I avoid this cost using lazy highlighting as you are aware.)

**Comment by monnier** [comment]
>> If the bug is fixed upstream it should be fixed directly in pcomplete.el.
> If by any chance you have both the time and the will, a fix there would be
> perfect. I've made the attempt to investigate this a couple of times in the
> past, but didn't manage to advance too far.

IIRC, it's indeed not trivial to fix.  It requires extricating the
"expand" functionality from the "complete" functionality, or something
like that.

As for upstream, I think `minibuffer.el` should start to discourage more
actively the use of the old calling convention.


        Stefan
